File: test/domain/entities/like_destination_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:like_spotify_mobile_app/domain/entities/like_destination.dart';

/// The pure half of the destination feature. `LikeDestination.kt` is the
/// Kotlin twin of this enum and `LikeDestinationTest.kt` mirrors these cases,
/// so the three like paths cannot drift apart.
void main() {
  group('legs', () {
    test('liked songs likes natively only', () {
      expect(LikeDestination.native.likesNatively, isTrue);
      expect(LikeDestination.native.addsToPlaylist, isFalse);
    });

    test('a playlist adds to the playlist only', () {
      expect(LikeDestination.playlist.likesNatively, isFalse);
      expect(LikeDestination.playlist.addsToPlaylist, isTrue);
    });

    test('both does both', () {
      expect(LikeDestination.both.likesNatively, isTrue);
      expect(LikeDestination.both.addsToPlaylist, isTrue);
    });
  });

  group('fromId', () {
    test('reads every id it writes', () {
      for (final destination in LikeDestination.values) {
        expect(LikeDestination.fromId(destination.id), destination);
      }
    });

    test('falls back to the default for a missing or unknown id', () {
      expect(LikeDestination.fromId(null), LikeDestination.native);
      expect(LikeDestination.fromId(''), LikeDestination.native);
      expect(LikeDestination.fromId('somewhere-new'), LikeDestination.native);
    });

    test('the default is the behaviour the app always had', () {
      expect(LikeDestination.defaultDestination, LikeDestination.native);
    });
  });

  group('resolve', () {
    test('keeps a destination that has the playlist name it needs', () {
      expect(
        LikeDestination.resolve(LikeDestination.playlist, 'Trigger likes'),
        LikeDestination.playlist,
      );
      expect(
        LikeDestination.resolve(LikeDestination.both, 'Trigger likes'),
        LikeDestination.both,
      );
    });

    test('degrades to liked songs when the name is missing', () {
      expect(
        LikeDestination.resolve(LikeDestination.playlist, '   '),
        LikeDestination.native,
      );
      expect(
        LikeDestination.resolve(LikeDestination.both, ''),
        LikeDestination.native,
      );
    });

    test('leaves liked songs alone whatever the name is', () {
      expect(
        LikeDestination.resolve(LikeDestination.native, ''),
        LikeDestination.native,
      );
      expect(
        LikeDestination.resolve(LikeDestination.native, 'Trigger likes'),
        LikeDestination.native,
      );
    });
  });
}

File: android/app/src/test/kotlin/com/osasuwu/like_spotify/LikeDestinationTest.kt
package com.osasuwu.like_spotify

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The destination branching both native like paths run on: `SpotifyLikeWorker`
 * asks [LikeDestination.likesNatively] / [LikeDestination.addsToPlaylist] which
 * legs to send and [LikeDestination.succeeded] whether the like counted, and
 * `YouTubeMusicLiker` asks exactly the same questions. The two paths cannot
 * drift while every decision lives here.
 *
 * The Kotlin twin of `test/domain/entities/like_destination_test.dart`; the
 * two files mirror each other on purpose.
 */
class LikeDestinationTest {

    // ---- which legs run ---------------------------------------------------

    @Test
    fun `liked songs likes natively only`() {
        assertTrue(LikeDestination.NATIVE.likesNatively)
        assertFalse(LikeDestination.NATIVE.addsToPlaylist)
    }

    @Test
    fun `a playlist adds to the playlist only`() {
        assertFalse(LikeDestination.PLAYLIST.likesNatively)
        assertTrue(LikeDestination.PLAYLIST.addsToPlaylist)
    }

    @Test
    fun `both does both`() {
        assertTrue(LikeDestination.BOTH.likesNatively)
        assertTrue(LikeDestination.BOTH.addsToPlaylist)
    }

    // ---- fromId -----------------------------------------------------------

    @Test
    fun `every id Dart writes reads back unchanged`() {
        // The ids are the Dart enum's, written into SharedPreferences by the
        // setRuleConfig channel call.
        assertEquals(LikeDestination.NATIVE, LikeDestination.fromId("native"))
        assertEquals(LikeDestination.PLAYLIST, LikeDestination.fromId("playlist"))
        assertEquals(LikeDestination.BOTH, LikeDestination.fromId("both"))
    }

    @Test
    fun `a missing or unknown id keeps the behaviour the app always had`() {
        assertEquals(LikeDestination.NATIVE, LikeDestination.DEFAULT)
        assertEquals(LikeDestination.DEFAULT, LikeDestination.fromId(null))
        assertEquals(LikeDestination.DEFAULT, LikeDestination.fromId(""))
        assertEquals(LikeDestination.DEFAULT, LikeDestination.fromId("somewhere-new"))
    }

    // ---- resolve ----------------------------------------------------------

    @Test
    fun `a named playlist destination is kept`() {
        assertEquals(LikeDestination.PLAYLIST, LikeDestination.resolve("playlist", "Trigger likes"))
        assertEquals(LikeDestination.BOTH, LikeDestination.resolve("both", "Trigger likes"))
    }

    @Test
    fun `a nameless playlist destination degrades to liked songs`() {
        // Nowhere to put the song: like it the old way rather than fail.
        assertEquals(LikeDestination.NATIVE, LikeDestination.resolve("playlist", ""))
        assertEquals(LikeDestination.NATIVE, LikeDestination.resolve("both", "   "))
    }

    @Test
    fun `liked songs ignores the playlist name`() {
        assertEquals(LikeDestination.NATIVE, LikeDestination.resolve("native", ""))
        assertEquals(LikeDestination.NATIVE, LikeDestination.resolve("native", "Trigger likes"))
    }

    // ---- succeeded --------------------------------------------------------

    @Test
    fun `a one-leg destination is exactly as good as its leg`() {
        assertTrue(LikeDestination.succeeded(LikeDestination.NATIVE, nativeOk = true, playlistOk = false))
        assertFalse(LikeDestination.succeeded(LikeDestination.NATIVE, nativeOk = false, playlistOk = true))
        assertTrue(LikeDestination.succeeded(LikeDestination.PLAYLIST, nativeOk = false, playlistOk = true))
        assertFalse(LikeDestination.succeeded(LikeDestination.PLAYLIST, nativeOk = true, playlistOk = false))
    }

    @Test
    fun `both fails only when both legs fail`() {
        assertTrue(LikeDestination.succeeded(LikeDestination.BOTH, nativeOk = true, playlistOk = true))
        assertTrue(LikeDestination.succeeded(LikeDestination.BOTH, nativeOk = true, playlistOk = false))
        assertTrue(LikeDestination.succeeded(LikeDestination.BOTH, nativeOk = false, playlistOk = true))
        assertFalse(LikeDestination.succeeded(LikeDestination.BOTH, nativeOk = false, playlistOk = false))
    }
}
